FormalConjectures/ForMathlib/Data/Set/Density.lean

```diff
@@ -25,6 +25,35 @@ open scoped Topology
 
 namespace Set
 
+/--
+Given a set `S` and an element `b` in an order `β`, where all intervals bounded above are finite,
+we define the partial density of `S` (relative to a set `A`) to be the proportion of elements in
+`{x ∈ A | x < b}` that lie in `S ∩ A`.
+
+This definition was inspired from https://github.com/b-mehta/unit-fractions
+-/
+noncomputable abbrev PartialDensity {β : Type*} [Preorder β] [LocallyFiniteOrderBot β]
+    (S : Set β) (A : Set β := Set.univ) (b : β) : ℝ :=
+  (S ∩ A ∩ Set.Iio b).ncard / (A ∩ Set.Iio b).ncard
+
+/--
+Given a set `S` in an order `β`, where all intervals bounded above are finite, we define the upper
+density of `S` (relative to a set `A`) to be the limsup of the partial densities of `S`
+(relative to `A`) for `b → ∞`.
+-/
+noncomputable def UpperDensity {β : Type*} [Preorder β] [LocallyFiniteOrderBot β]
+    (S : Set β) (A : Set β := Set.univ) : ℝ :=
+  atTop.limsup (fun (b : β) ↦ S.PartialDensity A b)
+
+/--
+Given a set `S` in an order `β`, where all intervals bounded above are finite, we define the upper
+density of `S` (relative to a set `A`) to be the liminf of the partial densities of `S`
+(relative to `A`) for `b → ∞`.
+-/
+noncomputable def LowerDensity {β : Type*} [Preorder β] [LocallyFiniteOrderBot β]
+    (S : Set β) (A : Set β := Set.univ) : ℝ :=
+  atTop.liminf (fun (b : β) ↦ S.PartialDensity A b)
```

Comment:
It would be nice to (eventually) have lemmas lowerDensity <= upperDensity, and lowerDensity = upperDensity implies HasDensity, but that can be todo for now :)